dace/transformation/interstate/if_extraction.py

```diff
@@ -0,0 +1,156 @@
+                    new_state_list.append(e.dst)
+            sdfg.remove_node(s)
+        state_list = new_state_list
+
+
+@make_properties
+class IfExtraction(transformation.MultiStateTransformation):
+    """
+    Detects an if statement as the root of a nested sdfg, and extracts it by computing it in the outer sdfg and
+    replicating the state containing the nested sdfg
+    """
+
+
+    root_state = transformation.PatternNode(sd.SDFGState)
+
+
+    @staticmethod
+    def annotates_memlets():
+        return False
+
+
+    @classmethod
+    def expressions(cls):
+        return [sdutil.node_path_graph(cls.root_state)]
+
+
+    def can_be_applied(self, graph, expr_index, sdfg, permissive=False):
+        root_state: SDFGState = self.root_state
+
+        out_edges = graph.out_edges(root_state)
+        in_edges = graph.in_edges(root_state)
+
+        if len(in_edges) > 0:
+            return False
+
+        if len(out_edges) != 2:
+            return False
+
+        # needs to be a nested sdfg
+        if not sdfg.parent:
+            return False
+
+        nested_sdfg: sd.nodes.NestedSDFG = sdfg.parent_nsdfg_node
+        parent_sdfg: sd.SDFG = sdfg.parent.sdfg
+
+        # collect outer symbols used in the interstate edges outgoing the if guard
+        if_symbols = set(str(nested) for e in out_edges for s in e.data.free_symbols
+                            for nested in symbolic.pystr_to_symbolic(nested_sdfg.symbol_mapping[s]).free_symbols)
+
+        # collect symbols available to state containing the nested sdfg
+        available_symbols = parent_sdfg.symbols.keys() | parent_sdfg.arglist().keys()
+        for desc in parent_sdfg.arrays.values():
+            available_symbols |= {str(s) for s in desc.free_symbols}
+
+        start_state = sdfg.start_state
+        for e in sdfg.predecessor_state_transitions(start_state):
+            available_symbols |= e.data.new_symbols(sdfg, available_symbols).keys()
+
+        # check if used symbols can be computed in the outer scope
+        if not if_symbols.issubset(available_symbols):
+            return False
+
+        # check if symbols are not written in the state containing the nested sdfg
+        _, wset = sdfg.parent.read_and_write_sets()
+        if len(if_symbols.intersection(wset)) != 0:
+            return False
+
+        return True
+
+
+    def apply(self, _, if_sdfg: sd.SDFG):
+        if_root_state: SDFGState = self.root_state
+        if_branch: SDFGState = if_sdfg.parent
+        outer_sdfg: sd.SDFG = if_branch.sdfg
+        if_nested_sdfg_node: sd.nodes.NestedSDFG = if_sdfg.parent_nsdfg_node
+
+        if_edge, else_edge = if_sdfg.out_edges(if_root_state)
+
+        # create new state to perform the if, and have it replace the state containing the nested SDFG
+        new_state = outer_sdfg.add_state()
+        sdutil.change_edge_dest(outer_sdfg, if_branch, new_state)
+
+        # take the old state as the if branch, and create a copy to act as the else branch
```

Comment:
I feel like "duplicate_element" should become a utility method of the sdfg

Reply:
How is this different from a deepcopy?

Reply:
It is already used in some other transformation because according to some comment "it is faster than deepcopy". Consider that deepcopy will copy the entire SDFG as it is referenced in the state. Deepcopying nodes can also lead to invalid state -> deepcopy shoud be used on very small things (edge data, ...) or avoided completely, and some state/node duplication utility should likely be provided in the state/node itself